### PR TITLE
Move minitest ,net-telnet, power_assert, rake, test-unit to bundled gems

### PR DESCRIPTION
**Pull request by bjfish**
Here is a summary of the changes. The two version updates are up to date with ruby-2.3.3

Removed from rake from stdlib
`./bin/ruby -rbundler-workarounds -S gem install rake -v 10.4.2 --no-doc`
Remove from stdlib, Upgrade minitest version: 5.4.1 -> 5.8.5	
`./bin/ruby -rbundler-workarounds -S gem install minitest -v 5.8.5 --no-doc`
Added power_assert version: 0.2.6
`./bin/ruby -rbundler-workarounds -S gem install power_assert -v 0.2.6 --no-doc`
Remove test-unit from stdlib, Upgrade version 3.1.1 -> 3.1.5	
`./bin/ruby -rbundler-workarounds -S gem install test-unit -v 3.1.5 --no-doc`
Remove net-telnet from stdlib
`./bin/ruby -rbundler-workarounds -S gem install net-telnet -v 0.1.1 --no-doc`
	
Add the above gems/gemspecs to gitignore
Add lib/bin/rake to gitignore

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -67,13 +67,23 @@ lib/ruby/gems/2.3.0/*
...
+!lib/ruby/gems/2.3.0/specifications/test-unit-3.1.5.gemspec
+!lib/ruby/gems/2.3.0/specifications/net-telnet-0.1.1.gemspec
 
 # Lib bin
 lib/bin/**
 !lib/bin/gem
-!lib/bin/irb
\ No newline at end of file
+!lib/bin/irb
+!lib/bin/rake
```
Comment (eregon) [inline .gitignore]: I take the habit to always keep a final `\n` for text files

File: lib/bin/rake [hunk 2]
```diff
@@ -0,0 +1,22 @@
+#!/Users/brandonfish/Documents/jruby/lib/bin/truffleruby
```
Comment (eregon) [inline lib/bin/rake]: This path is hardcoded, how can we fix that?

File: .gitignore [hunk 3]
```diff
@@ -67,13 +67,23 @@ lib/ruby/gems/2.3.0/*
 !lib/ruby/gems/2.3.0/gems
 lib/ruby/gems/2.3.0/gems/*
 !lib/ruby/gems/2.3.0/gems/did_you_mean-1.0.0
+!lib/ruby/gems/2.3.0/gems/rake-10.4.2
+!lib/ruby/gems/2.3.0/gems/minitest-5.8.5
+!lib/ruby/gems/2.3.0/gems/power_assert-0.2.6
+!lib/ruby/gems/2.3.0/gems/test-unit-3.1.5
+!lib/ruby/gems/2.3.0/gems/net-telnet-0.1.1
```
Comment (eregon) [inline .gitignore]: I would add a new line after this so gems and specifications are two "groups"

**Comment by eregon** [comment]
Great work :+1:

**Comment by chrisseaton** [comment]
Can you document how to upgrade these gems in `doc/contributing/upgrading-ruby.md`, including how to work out from MRI what version to upgrade them to.

Do we need to update what is in the root `bin` directory.

File: lib/bin/rake (hunk 2 above)
Comment (bjfish) [inline lib/bin/rake]: The `bin` and `lib/bin` setup still confuses me a little. In `rbenv`, For 2.3.3 I see `#!/usr/bin/env ruby_executable_hooks` and `rbenv` jruby `#!/Users/brandonfish/.rbenv/versions/jruby-9.1.7.0/bin/jruby`.  I believe (not 100% sure) `rbenv` updates these via the `rbenv rehash` command: https://github.com/rbenv/rbenv#rbenv-rehash 
https://github.com/rbenv/rbenv/blob/d0779fc8fb90fb7268d0ee858086e92bcd316787/rbenv.d/exec/gem-rehash/rubygems_plugin.rb#L1

My local ruby dev build adds `#!/Users/brandonfish/.rbenv/versions/ruby-dev/bin/ruby`. I think jruby avoided any potential conflicts by using `#!/usr/bin/env jruby` in there repo but this implies `jruby` is on the path. Maybe the safest this to do for now would be `#!/usr/bin/env truffleruby`.  I'll keep investigating the bin files to see if I can learn more.

If we want to get these correct, it seems like they should be updated at install time.

Comment (chrisseaton) [inline lib/bin/rake]: Everything in `lib/bin` should be a Ruby script, designed to run with `-S`. `bin` should shell scripts that perhaps run something from `lib/bin` with `-S`. The reason for this is that if `bin` files are Ruby files with shebangs, which Ruby would they run? They rely on the Ruby you want being `env ruby`.

File: .gitignore (hunk 1 above)
Comment (bjfish) [inline .gitignore]: I've added the final `\n`

File: .gitignore (hunk 3 above)
Comment (bjfish) [inline .gitignore]: Added a new line here

**Comment by bjfish** [comment]
@chrisseaton I've added the instructions on updating bundled gems.

File: doc/contributor/updating-ruby.md [hunk 4]
```diff
@@ -14,3 +14,13 @@ questions.
...
+
+## Updating bundled gems
+The current list of bundled gems their versions are found here:
+https://github.com/ruby/ruby/blob/ruby_2_3/gems/bundled_gems
+
+To update a bundled gem, follow these steps:
+1. Remove the current gem and gemspec from `lib/gems/2.3.0/gems` and `lib/gems/2.3.0/specifications`
+2. Run the gem install command with the desired version. E.g. `./bin/ruby -rbundler-workarounds -S gem install rake -v 10.4.2 --no-doc`
+3. Update the project `.gitignore` to allow the newly install gem sources and gemspec
+4. If the gem installs any executables like `rake`. Add these to the `.gitignore` if not already and verify there is no shebang line and remove executable permissions on the file.
```
Comment (eregon) [inline doc/contributor/updating-ruby.md]: The list formatting did not work out (use the `rich diff` button above). Maybe an extra space before `1.`.
An not-numbered list might be more consistent with above.

**Comment by eregon** [comment]
I think `bin/rake` should be changed to be like `bin/irb` now.
(Actually all launchers in `bin/` should be changed like that I think)

File: doc/contributor/updating-ruby.md (hunk 4 above)
Comment (bjfish) [inline doc/contributor/updating-ruby.md]: Thanks, fixed it.

**Comment by chrisseaton** [comment]
Yes everything in `bin` except for `truffleruby`, `ruby` and `jruby-truffle-tool` should look like

```bash
#!/usr/bin/env bash
exec "$(dirname $0)/truffleruby" -Sirb "$@"
```

**Comment by bjfish** [comment]
I've fixed the `bin/rake` file

**Comment by bjfish** [comment]
The failing test appears to work for me locally. I even did a fresh checkout of this branch to try it. Maybe it's a Linux thing?  Or maybe just comment this out for now? I can investigate more.

**Comment by bjfish** [comment]
@eregon Maybe I need to use the `:unsetenv_others` option for system

**Comment by eregon** [comment]
@bjfish You need t swap the 2 lines in `before_install`, GEM_PATH should be unset after switching with `rvm`. It only shows how brittle we are to run when GEM_PATH or GEM_HOME is set though :cry:

**Comment by bjfish** [comment]
@eregon Thanks, trying it now.

The failure may be a result of pointing to a different ruby version's GEM_HOME. It looks like 2.1 had different default gems. So, this failure might not be brittleness.

**Comment by eregon** [comment]
> The failure may be a result of pointing to a different ruby version's GEM_HOME/PATH. It looks like 2.1 had different default gems. So, this failure might not be brittleness.

It is brittle in that running bin/truffleruby with some random GEM_HOME set by e.g. `rvm` does not find even its own bundled gems!
Using another Ruby's gems might lead to all sort of problems as well, even if it works at first.

**Comment by bjfish** [comment]
@eregon With ruby 2.3 I get the same error. It looks like setting GEM_HOME and GEM_PATH will completely override the default_gem_dir
```
$ GEM_PATH=/Users/brandonfish/.rbenv/versions/2.1.10/lib/ruby/gems/2.1.10 GEM_HOME=/Users/brandonfish/.rbenv/versions/2.1.10/lib/ruby/gems/2.1.10 ruby -e 'puts gem("minitest")'
/Users/brandonfish/.rbenv/versions/2.3.0/lib/ruby/2.3.0/rubygems/dependency.rb:319:in `to_specs': Could not find 'minitest' (>= 0) among 5 total gem(s) (Gem::LoadError)
Checked in 'GEM_PATH=/Users/brandonfish/.rbenv/versions/2.1.10/lib/ruby/gems/2.1.10', execute `gem env` for more information
	from /Users/brandonfish/.rbenv/versions/2.3.0/lib/ruby/2.3.0/rubygems/dependency.rb:328:in `to_spec'
	from /Users/brandonfish/.rbenv/versions/2.3.0/lib/ruby/2.3.0/rubygems/core_ext/kernel_gem.rb:65:in `gem'
	from -e:1:in `<main>'

```

**Pull request merged**